### ISSUE TITLE
SCR commands in Slurm script

### ISSUE THREAD

**Issue by gongotar** [LLNL/scr#93]
Hi,

I'm currently trying to get scr work in slurm script. I have used the example script from the tutorial. but I cannot call scr commands in the script. I get following errors by running sbatch slurm_script:

/local/tmp/slurmd/job08249/slurm_script: line 8: /usr/local/tools/dotkit/init.sh: File or directory not found
/local/tmp/slurmd/job08249/slurm_script: line 9: use: Command not found
/local/tmp/slurmd/job08249/slurm_script: line 23: scr_srun: Command not found

In fact there is no init.sh or dotkit installed on my system. I couldn't find anything in the source nor in the documentation.

**Comment by adammoody** [comment]
There are some LLNL-specific items in that example.  You can ignore the dotkit init.sh and use commands.

However, you should probably have an scr_srun if SCR installed properly.  Do you see that?

**Comment by gongotar** [comment]
No, i cannot add scr commands to the script. As you can see in the errors, I get an error for scr_srun command. This is the script I'm trying to execute:

`#!/bin/bash
#MSUB -l partition=atlas
#MSUB -l nodes=2
#MSUB -l resfailpolicy=ignore
# above, tell MOAB / SLURM to not kill the job allocation upon a node failure
# also note that the job requested 2 spares -- it uses 64 nodes but allocated 66
# add the scr commands to the job environment
. /usr/local/tools/dotkit/init.sh
use scr-1.1
# specify where checkpoint directories should be written
export SCR_PREFIX=/nfs/csr/bzcghola/scr/directories/
# instruct SCR to flush to the file system every -1 checkpoints
export SCR_FLUSH=-1
# halt if there is less than an hour remaining (3600 seconds)
export SCR_HALT_SECONDS=3600
# attempt to run the job up to 3 times
export SCR_RUNS=3
# set the job id
SCR_JOB_ID=456
# set username
SCR_USER_NAME=bzcghola
# run the job with scr_srun
scr_srun -n1 ./st_test_scr
`

SCR has been installed correctly I think! I can use it without slurm. Just by testing some sample codes using scr to write checkpoints. Is there any thing I should to in order to make slurm recognise SCR?

**Comment by gongotar** [comment]
Hi again, 
I have modified my script as follows:
```
#!/bin/bash
#SBATCH --nodelist=cumu01-00
export SCR_FLUSH=-1
export SCR_HALT_SECONDS=3600
export SCR_RUNS=3
export SCR_PREFIX=<prefix address>
export SCR_JOB_ID=<job id>
export SCR_USER_NAME=<username>
export SCR_CLUSTER_NAME=<clustername>
export SCR_JOB_NAME=<jobname>
scr_srun -n 1 my_scr_program
```

after running the slurm script I can see scr tries to ping the host cumu01 which is not a node name, so there is no host with the name 'cumu01'. It is actually the partition name containing nodes cumu01-[00-15]. This means scr is confused probably in defining the node name and should ping the host cumu01-00 (which is available). This problem occurs also if I don't define a node list (line 2) or if I use more than one node (scr-srun -n 2). This is the output I get from scr_srun:

> scr_srun: Started: Mo 23. Okt 16:40:06 CEST 2017
> scr_srun: prerun: Mo 23. Okt 16:40:07 CEST 2017
> scr_prerun: Started: Mo 23. Okt 16:40:07 CEST 2017
> scr_prerun: Ended: Mo 23. Okt 16:40:07 CEST 2017
> scr_prerun: secs: 0
> scr_prerun: exit code: 0
> ping: unknown host cumu01
> ping: unknown host cumu01
> cumu01: Failed to ping
> SCR: Failed node detected: JOBID=8747 ATTEMPT=0 TIME=1508769607 NNODES=-1 RUNTIME=0 FAILED=cumu[01]
> scr_srun: (Nodes remaining=0) < (Nodes needed=1), ending run.
> scr_srun: postrun: Mo 23. Okt 16:40:07 CEST 2017
> scr_postrun: Started: Mo 23. Okt 16:40:07 CEST 2017
> ping: unknown host cumu01
> scr_postrun: UPNODES:
> scr_postrun: Ended: Mo 23. Okt 16:40:07 CEST 2017
> scr_postrun: secs: 0
> scr_postrun: exit code: 1
> scr_srun: ERROR: Command failed: scr_postrun -p <scr_prefix>
> scr_srun: Ended: Mo 23. Okt 16:40:07 CEST 2017

In addition by changing the wrapper script a little bit to use two nodes (cumu01-00,cumu01-01) in partition _cumu01_ as follows I get another error from scr_srun:

```
#!/bin/bash
#SBATCH --nodelist=cumu01-00,cumu01-01
export SCR_FLUSH=-1
export SCR_HALT_SECONDS=3600
export SCR_RUNS=3
export SCR_PREFIX=<prefix address>
export SCR_JOB_ID=<job id>
export SCR_USER_NAME=<username>
export SCR_CLUSTER_NAME=<clustername>
export SCR_JOB_NAME=<jobname>
scr_srun -p cumu01 -n 2 -N 2 my_scr_program
```

scr_srun error:

> scr_srun: Started: Mo 23. Okt 16:22:00 CEST 2017
scr_srun: ERROR: scr_run is executing on cumu01-00, which is not part of the job's nodeset cumu01-[00-01].

Do you know whats happening and how can I make SCR recognise my partition and nodes structure correctly?

**Comment by adammoody** [comment]
Yes, we have some more LLNL-specific assumptions built into the node name formats.  We're working to remove those in future releases, but they're in the version you're using here.  For now, try falling back to using "srun" instead of "scr_srun".  That should run the app using the library at least.  Please give that a shot and let me know if that works.

**Comment by adammoody** [comment]
The master branch has been updated with a commit that may help here:
https://github.com/LLNL/scr/commit/6bb59ac67225ca400d0e86ae82062c4e7c6c100f

**Comment by gongotar** [comment]
Thanks for the commit, it works fine now. Using scr-1.1.8 I couldn't use _scr_srun_ however _srun_ has worked without any problems, now by using the master branch I can call _scr_srun_ to run my jobs. However I get lots of warnings from it. They don't prevent the execution of jobs but I put them here anyway just in case you need them for debugging proposes:

> Unescaped left brace in regex is deprecated, passed through in regex; marked by <-- HERE in m/(\$\w+)|(\${ <-- HERE \w+})/ at \<SCR_HOME\>/share/scr/scr_param.pm line 120.
Unescaped left brace in regex is deprecated, passed through in regex; marked by <-- HERE in m/\${ <-- HERE ?(\w+)}?/ at \<SCR_HOME\>/share/scr/scr_param.pm line 124.

There was actually a lot of warnings but i've just printed only one version of each of the message types in the output.
Another issue I experienced by using _scr_srun_ of the master version is that the allocation doesn't expire automatically after the job is finished and `SCR_Finalize` is called successfully. As you can see I get messages which indicate the successful execution of my job.



> \<job output and results will be printed here\>
> scr_retries_halt: HALT RUN: Reason: SCR_FINALIZE_CALLED.
> scr_srun: Halt condition detected, ending run.
> scr_srun: postrun: Mi 25. Okt 12:59:28 CEST 2017
> scr_postrun: Started: Mi 25. Okt 12:59:28 CEST 2017
> scr_postrun: UPNODES:   cumu02-14
> scr_postrun: Looking for output sets
> scr_postrun: Found no output set to scavenge
> scr_postrun: Looking for most recent checkpoint
> scr_postrun: Attempting to scavenge checkpoint dataset 10
> scr_postrun: Scavenging files from cache for checkpoint scr.dataset.10 to \<SCR_PREFIX\>/.scr/scr.dataset.10
> scr_postrun: \<SCR_HOME\>/bin/scr_scavenge --id 10 --from \<tmp-dir\> --to \<SCR_PREFIX\> --jobset cumu02-14 --up cumu02-14
> scr_scavenge: Wed Oct 25 12:59:28 2017
> scr_scavenge: \<PDSH_HOME\>/bin/pdsh -Rexec -f 256 -S -w 'cumu02-14' srun -n1 -N1 -w %h \<SCR_HOME\>/bin/scr_copy --cntldir <tmp-dir> --id 10 --prefix \<SCR_PREFIX\> --buf 1048576 --crc  --containers




In other words I can see that my code is executed and finished successfully but the allocation is still there and I should manually call `scancel <jobid>` in order to terminate it. This isn't the case if I use _srun_ instead of _scr_srun_. Have I missed something in my script or my code?

**Comment by adammoody** [comment]
For the first problem about the perl warnings, it looks like you have a newer/stricter version of perl that I do.  We can fix those warnings.  What output do you get from "perl --version"?

**Comment by adammoody** [comment]
Do you have regex warnings like above from lines other than scr_param.pm line 120 and 124?

**Comment by gongotar** [comment]
The output of perl --version is as follows:
> This is perl 5, version 22, subversion 1 (v5.22.1) 

The warnings are repeated many times for scr_param.pm lines 120 and 124. There are no other lines in the warnings.

**Comment by adammoody** [comment]
Ok, thanks.  I think I have a fix in the pipeline for the scr_param.pm warnings.

For the hang, I see the last line printed is using pdsh to run scr_copy on the the node.  Perhaps that command is hanging.  What happens if you manually execute that command?  Does it hang or complete?

**Comment by gongotar** [comment]
I have executed the _pdsh_ command in the last line after the job finished (as said before, resources are still allocated) and it hangs without any errors or outputs.

**Comment by adammoody** [comment]
We've got a commit on master now that I think should fix the scr_param.pm warning messages.  Please give that a shot and let me know if it helps.

I'm still puzzled about the hang with scr_srun, so we still need to get to the bottom of that.  If you login into the job node while it's hanging, can you see any processes listed under ps?

**Comment by gongotar** [comment]
I have tested the master version and the warnings are gone now. Thanks for the commit.
After running the job and hanging _scr_srun_, running command `ps` in the job node delivers following processes:

> 115042 ?        00:00:00 slurm_script
115050 ?        00:00:00 scr_srun
115248 ?        00:00:00 scr_postrun
115304 ?        00:00:00 scr_scavenge
115309 ?        00:00:00 sh
115310 ?        00:00:00 pdsh
115314 ?        00:00:00 srun
115315 ?        00:00:00 srun
115397 ?        00:00:00 ps

**Comment by adammoody** [comment]
This suggests to me that perhaps the pdsh in the scavenge operation is stuck.  The calling sequence here is:

scr_srun --> scr_postrun --> scr_scavenge --> pdsh --> srun

Let me look at that to get some ideas.

**Comment by gongotar** [comment]
Write me here if you need more information about the hang. Maybe I can run it in debug mode or something like that if it helps.

**Comment by gongotar** [comment]
s
